File: haskell/problems/1-10/1/P1.hs
module P1 where

{-
If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.
Find the sum of all the multiples of 3 or 5 below 1000.
-}

main = print $ sumMultiplesBelow1000 3 + sumMultiplesBelow1000 5 - sumMultiplesBelow1000 (lcm 3 5)

sumMultiplesBelow1000 :: Integer -> Integer
sumMultiplesBelow1000 n = n * sumOneToN (div 999 n)

sumOneToN :: Integer -> Integer
sumOneToN n = n * (n+1) `div` 2

lcm' :: Integer -> Integer -> Integer
lcm' x y = div (x * y) (gcf' x y)

gcf' :: Integer -> Integer -> Integer
gcf' x y
  | x < y = gcf' y x
  | mod x y == 0 = y
  | otherwise = gcf' y (mod x y)

File: c++/problems/1-10/1/P1.cpp

#include <iostream>
using namespace std;

int sumOfMultiplesLessThan1000(int);
int sumOneToN(int);
int lcm(int, int);
int gcd(int, int);

int main() {
  int ans = sumOfMultiplesLessThan1000(3)
          + sumOfMultiplesLessThan1000(5)
          - sumOfMultiplesLessThan1000(lcm(3,5));
  cout << ans;
  return 0;
}

int sumOfMultiplesLessThan1000(int n) {
  return n * sumOneToN(999 / n);
}

int sumOneToN(int n) {
  return n * (n+1) / 2;
}

int lcm(int x, int y) {
  return x*y / gcd(x,y);
}

int gcd(int x, int y) {
  if (x < y) {
    return gcd(y, x);
  }
  int m = x % y;
  if (m == 0) {
    return y;
  }
  return gcd(m, y);
}



